components/tikv_util/src/logger/file_log.rs

```diff
@@ -133,64 +147,169 @@ impl Drop for RotatingFileLogger {
     }
 }
 
+pub struct RotateByTime {
+    rotation_timespan: Duration,
+    next_rotation_time: DateTime<Utc>,
+}
+
+impl RotateByTime {
+    pub fn new(rotation_timespan: Duration) -> Self {
+        Self {
+            rotation_timespan,
+            next_rotation_time: compute_rotation_time(&Utc::now(), rotation_timespan),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_for_test(next_rotation_time: DateTime<Utc>, rotation_timespan: Duration) -> Self {
+        Self {
+            next_rotation_time,
+            rotation_timespan,
+        }
+    }
+}
+
+impl Rotator for RotateByTime {
+    fn should_rotate(&self, _: &File) -> io::Result<bool> {
+        Ok(Utc::now() > self.next_rotation_time)
+    }
+
+    fn rotate(&mut self, path: &Path, file: &mut File) -> io::Result<Option<PathBuf>> {
+        file.flush()?;
+
+        let renamed = match rename_by_timestamp(path) {
+            Ok(path) => Some(path),
+            Err(ref e) if e.kind() == ErrorKind::NotFound => None,
+            Err(e) => return Err(e),
+        };
+
+        self.next_rotation_time = compute_rotation_time(&Utc::now(), self.rotation_timespan);
+
+        *file = open_log_file(&path)?;
+
+        Ok(renamed)
+    }
+
+    fn is_valid(&self) -> bool {
+        !self.rotation_timespan.is_zero()
+    }
+}
+
+pub struct RotateBySize {
+    rotation_size: ReadableSize,
+}
+
+impl RotateBySize {
+    pub fn new(rotation_size: ReadableSize) -> Self {
+        RotateBySize { rotation_size }
+    }
+}
+
+impl Rotator for RotateBySize {
+    fn should_rotate(&self, file: &File) -> io::Result<bool> {
+        Ok(file.metadata()?.len() > self.rotation_size.as_b())
+    }
+
+    fn rotate(&mut self, path: &Path, file: &mut File) -> io::Result<Option<PathBuf>> {
+        file.flush()?;
+
+        let renamed = match rename_by_timestamp(path) {
+            Ok(path) => Some(path),
+            Err(ref e) if e.kind() == ErrorKind::NotFound => None,
+            Err(e) => return Err(e),
+        };
+
+        *file = open_log_file(&path)?;
+
+        Ok(renamed)
+    }
+
+    fn is_valid(&self) -> bool {
+        self.rotation_size.as_b() != 0
+    }
+}
+
 #[cfg(test)]
 mod tests {
-    use std::fs::OpenOptions;
-    use std::io::prelude::*;
-    use std::path::Path;
+    use super::*;
 
     use chrono::{Duration, Utc};
+    use std::sync::Mutex;
     use tempfile::TempDir;
-    use utime;
 
-    use super::RotatingFileLogger;
+    // Workaround for `failpoint`
```

Comment:
Suggestion: FIXME with the fallpoint-rs issue link

Reply:
Done, use `FailScenario` instead.